### PR TITLE
Utkarsh/issue 186

### PR DESCRIPTION
**Pull request by Sigma-Verma**
Made cache removal safer by using a lock to avoid repeated checks

# Pull Request

## Description

Enhanced thread safety by consolidating cache removal logic within a lock, preventing redundant iterations and ensuring consistent access to shared data.

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by peterdudfield — CHANGES_REQUESTED** [review]
Thanks @Sigma-Verma for this. 

Do you mind fixing the linting, see CI?

This seems a really good way to do it, with thread locking. Do you know how this works if two calls to the api are made very close to each other? Does this then slow down the second one? Perhaps this is the only way to do it?

Perhaps there's a speed up here, where we check if these items are cache locked, and if so carry on and do nothing

**Comment by Sigma-Verma** [comment]
hey can you guide me on how to fix the Linting ?
and for the question about two very close calls, 
i added a catch entry removal check
If no entries need to be removed , it will skip the locking process

**Comment by peterdudfield** [comment]
> hey can you guide me on how to fix the Linting ? and for the question about two very close calls, i added a catch entry removal check If no entries need to be removed , it will skip the locking process

See errors in the github actinos and fix manually
Screenshot 2024-09-30 at 23 01 48
or do `make format`

File: pv_site_api/cache.py [hunk 1]
```diff
@@ -17,33 +17,47 @@
...
-
-    for key in keys_to_remove:
-        last_updated.pop(key)
-        response.pop(key)
-
-    process = psutil.Process(os.getpid())
-    logger.debug(f"Memory is {process.memory_info().rss / 10 ** 6} MB")
-
-    return last_updated, response
+        now = datetime.now(tz=timezone.utc)
```
Comment (peterdudfield) [inline pv_site_api/cache.py:35]: is this identented correctly?

**Comment by peterdudfield** [comment]
Also, do you mind incorporating #199?

File: pv_site_api/cache.py [hunk 2]
```diff
@@ -28,18 +31,28 @@ def remove_old_cache(
...
+
     keys_to_remove = []
-    last_updated_copy = last_updated.copy()
-    for key, value in last_updated_copy.items():
-        if now - timedelta(seconds=remove_cache_time_seconds) > value:
-            logger.debug(f"Removing {key} from cache, ({value})")
-            keys_to_remove.append(key)
+
+    with cache_lock:
+        for key, value in last_updated.items():
```
Comment (peterdudfield) [inline pv_site_api/cache.py:48]: could you include the .copy() that i added please

**Comment by peterdudfield** [comment]
Thank you for this, just a small lint change, 

Screenshot 2024-10-18 at 10 29 08

**Pull request merged**